draft-irtf-nwcrg-coding-and-congestion-01.txt

```diff
@@ -1,504 +1,528 @@
+
+Internet-Draft           Coding and congestion              January 2020
+
+
+                  |                  --------                  ^
+                  | data             |      | data,            | sending
+                  |                  |      | requirements     | rate (or
+                  v                  |      v                  | window)
+   +-----------------------------+   |    +-------------------------+
+   |             FEC             | <----> |           CC            |
+   +-----------------------------+   |    +-------------------------+
+     ^                   |           |              ^
+     | information       | coded     |              | network
+     | about repaired    | data      |              | measurements
+     | blocks            ------------               |
+
+   As this diagram shows, the inputs to FEC (data to work upon, and
+   signaling from the receiver about losses and/or repaired blocks) are
+   distinct from the inputs to CC. The latter calculates a sending rate
+   or window from network measurements, and it takes the data to send as
+   input, sometimes along with application requirements such as upper/
+   lower rate bounds, periods of quiescence, or a priority.
+
+   It is not clear that the ACK signals feeding into a congestion
+   control algorithm are useful to FEC in their raw form, and vice versa
+   - information about repaired blocks may be quite irrelevant to a CC
+   algorithm.  However, there can be meaningful other interactions
+   (indicated by the horizontal double arrow) between the two entities,
+   usually as a result of their operation rather than by relaying their
+   own raw inputs.  For example, the network measurements carried out by
+   CC can yield a longer-term statistical measure such as a loss ratio
+   which is useful input for a FEC coding scheme.  Similarly, unequal
+   error protection using fountain codes can be used to assign different
+   priorities to blocks of data, and these priorities can be honored by
+   a CC mechanism.
+
+3.  Scope
+
+   This section describes the scope of the document.
+
+3.1.  Type of application
+
+   The document focuses on reliable transmissions.
+
+3.2.  End-to-end
+
+   The document focuses on end-to-end coding, i.e.  cases where coding
+   is added at the server and client end points.  The discussions should
+   then consider fairness with non-coding solutions.
+
+4.  Base solution description
+
+   The solution can be described as follows: [MICHAEL: So far, it hasn't
+   been clear that we're going to describe a "solution".  A solution to
+   which problem?  I thought we're only discussing how FEC and CC
+   relate.]
+
+
+Kuhn, Lochin, Michel & WeExpires July 11, 2020                  [Page 4]
+
+Internet-Draft           Coding and congestion              January 2020
+
+
+   o  The client indicates to the server that one or multiple packets
+      have been repaired using a coding scheme.  The "repaired packet"

```

Comment:
We chose to force to have a "repaired" signal here because in some cases (such as TCP), it might be complicated in practice to not hide the loss signal and avoid a retransmission at the same time. With a coded tunnel under TCP, providing the "repaired". But I agree this idea looks a bit weird. 

Maybe we need another solution : 

Instead of forcing the sending of a "repaired" signal, we could distinguish two cases : 
- if the coding is performed in a layer between IP and congestion-controlled transport protocol, the coding layer must include a mechanism to detect network congestion and when congestion seems to be present, it must relay the information to the above layer using mechanisms such as ECN
- if the coding is performed as part of the congestion-controlled transport protocol, if a packet is lost + recovered, the receiver must not hide the loss signal of the packet to the sender.

Does this look better ? If you think it is a good idea, I can try to integrate it into the document.

Reply:
This matches what I just wrote in the discussion with @elochin here: https://github.com/irtf-nwcrg/draft-irtf-nwcrg-coding-and-congestion-in-transport/issues/3   we should discuss implementation strategies, but organize them by layering. This will be a good framework for statements like the ones you wrote above (in the bullets).

I suggest: I do an update where I remove section 6 and 7, start with a section on implementation discussing layering, and you take it from there.